hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/partitioner/BucketIndexPartitioner.java

```diff
@@ -33,22 +38,36 @@
  */
 public class BucketIndexPartitioner<T extends HoodieKey> implements Partitioner<T> {
 
-  private final int bucketNum;
+  private final Configuration conf;
   private final String indexKeyFields;
+  private final String hashingInstantToLoad;
+  private StorageConfiguration<org.apache.hadoop.conf.Configuration> storageConf;
 
-  private Functions.Function2<String, Integer, Integer> partitionIndexFunc;
+  private Functions.Function3<Integer, String, Integer, Integer> partitionIndexFunc;
 
-  public BucketIndexPartitioner(int bucketNum, String indexKeyFields) {
-    this.bucketNum = bucketNum;
+  public BucketIndexPartitioner(Configuration conf, String indexKeyFields,
+                                StorageConfiguration<org.apache.hadoop.conf.Configuration> storageConf) {
+    this.conf = conf;
     this.indexKeyFields = indexKeyFields;
+    this.hashingInstantToLoad = conf.get(FlinkOptions.BUCKET_INDEX_PARTITION_LOAD_INSTANT);
+    this.storageConf = storageConf;
   }
 
   @Override
   public int partition(HoodieKey key, int numPartitions) {
     if (this.partitionIndexFunc == null) {
-      this.partitionIndexFunc = BucketIndexUtil.getPartitionIndexFunc(bucketNum, numPartitions);
+      this.partitionIndexFunc = BucketIndexUtil.getPartitionIndexFunc(numPartitions);
+    }
+
+    int bucketNum;
+    if (!StringUtils.isNullOrEmpty(hashingInstantToLoad)) {
+      org.apache.hadoop.conf.Configuration hadoopConf = storageConf.unwrapAs(org.apache.hadoop.conf.Configuration.class);
+      PartitionBucketIndexCalculator calc = PartitionBucketIndexCalculator.getInstance(hashingInstantToLoad, hadoopConf, conf.get(FlinkOptions.PATH));
+      bucketNum = calc.computeNumBuckets(key.getPartitionPath());
+    } else {
+      bucketNum = conf.get(FlinkOptions.BUCKET_INDEX_NUM_BUCKETS);
     }
     int curBucket = BucketIdentifier.getBucketId(key.getRecordKey(), indexKeyFields, bucketNum);
-    return this.partitionIndexFunc.apply(key.getPartitionPath(), curBucket);
+    return this.partitionIndexFunc.apply(bucketNum, key.getPartitionPath(), curBucket);
```

Comment:
Can we fire a new partioner or we add a func `NumBucketsFunction`.

Reply:
add a new function NumBucketsFunction